Enable to check pre-release updates

hevrard opened the pull request:
Nightly is only checked if pre-releases are checked, which is
currently a hardcoded "false" in the UI. Next step is to enable
checking pre-releases from the UI.

AWoloszyn commented:
Until we actually have the builds coming out, we should probably hold on submitting this. It does look fine, but it will be dead code until the builds are coming out on a regular basis.

hevrard commented:
I revisited this based on feedback.

> * "nightly" has a specific meaning and makes a promise we are not going to keep

Now they are just "preview release"

> * tags should be static and not move

Fixed, with version tags now supporting an optional "preview" field which is expected to be the date in YYYYMMDD format
`v<major>.<minor>.<point>-<preview>`, with `-<preview>` being optional.
e.g. `v1.7.0.20190831`

> * once a release is public, I don't think it should ever change, even if it's marked as a pre-release

So now the tags will not move. We will just clean-up old pre-releases (assets etc) from GitHub once per quarter. Tags will stay around.

@pmuetschard I would need your advice on how best to trigger a UpdateWatcher.doCheck()  whenever the settings are modified?
Right now updates seem to be checked only every 8 hours, or at the very start if no `~/.gapic` file is found. It does not even wait for the WelcomeDialog to be completed before checking if there is an update. When a user goes to setting where he may change the update settings, I feel like we should trigger an update check. Yet I'm a bit confused on where to put that logic -- I don't have much experience in UI programming. Any advice much appreciated!

AWoloszyn commented:
I left a comment elsewhere, but v1.7.0-20190831 is the wrong format for the tags. 
Specifically because that is the tag BEFORE we release 1.7.0.  That is very misleading.

pmuetschard commented on core/app/run.go:
```diff
@@ -70,6 +70,8 @@ type VersionSpec struct {
 	// Point version, not used if <0
 	Point int
+	// Dev version, not used if <0
+	Dev int
```
Shouldn't this just be part of the build string?

hevrard replied:
The point here is to be able to use this field in `func (v VersionSpec) GreaterThan(o VersionSpec) bool`

If we don't store this "dev" part of the version as a field, then the whole version comparison logic becomes convoluted as we would use `GreaterThan()` only for the regular part of the version, and then compare the "dev" version on the side.

Hence the addition of the "dev" version as a field.

hevrard replied:
I got rid adding the "Dev" field to the Version

pmuetschard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -156,6 +162,15 @@ protected void configureDialog(FileDialog dialog) {
       adbListener.handleEvent(null);
+
+      allowUpdateChecks.addSelectionListener(new SelectionAdapter()
+        {
```
braces don't get their own lines in the Google Java Style guide.

hevrard replied:
thanks, done

pmuetschard commented on core/app/run.go:
```diff
@@ -97,6 +97,11 @@ func (v VersionSpec) GreaterThan(o VersionSpec) bool {
 	}
 }
 
+// Equal returns true if v is equal to o.
```
Since this ignores the Build field, the comment should be updated to point that out.

hevrard replied:
done

pmuetschard commented on gapis/server/server.go:
```diff
@@ -129,13 +129,27 @@ func (s *server) CheckForUpdates(ctx context.Context, includePrereleases bool) (
+		numFields, err := fmt.Sscanf(tagName, "v%d.%d.%d-dev-%d", &version.Major, &version.Minor, &version.Point, &devVersion)
+		if err != nil && numFields < 3 {
+			// some non-release tags have other format, e.g. libinterceptor-v1.0
+			log.W(ctx, "Ignoring tag %s", tagName)
```
This should not be a warning.

hevrard replied:
Changed to an Info

pmuetschard commented on gapis/server/server.go:
```diff
@@ -129,13 +129,27 @@ func (s *server) CheckForUpdates(ctx context.Context, includePrereleases bool) (
+			continue
+		}
+
+		if version.GreaterThan(mostRecentVersion) || (version.Equal(mostRecentVersion) && devVersion > mostRecentDevVersion) {
```
Aren't dev releases going to be using the next version's major.minor? If that is the case, this is incorrect: If we currently are on M.N, and the dev release is M.(N+1), when we release M.(N+1), it will not be considered newer than the last M.(N+1) dev release.

hevrard replied:
Fixed

pmuetschard commented on gapic/src/main/com/google/gapid/models/Settings.java:
```diff
@@ -88,6 +88,7 @@
   public String[] recentFiles = new String[0];
   public String adb = "";
   public boolean autoCheckForUpdates = true;
+  public boolean autoCheckForUpdatePrereleases = false;
```
autoCheckForUpdatePrereleases -> onDevTrack
updates.autoCheckPrerelease -> updates.devTrack

hevrard replied:
I renamed to "includePrerelease" to keep the "prerelease" naming because:
- it reflects the GitHub naming
- it is already used in other parts of GAPIC, e.g. in function `checkForUpdates(boolean includePrereleases)`

pmuetschard commented on gapic/src/main/com/google/gapid/util/Messages.java:
```diff
@@ -60,6 +60,7 @@
   public static final String CRASH_REPORTING_OPTION =
       "Help GAPID identify issues by sending crash reports to Google";
   public static final String UPDATE_CHECK_OPTION = "Automatically check for GAPID updates";
+  public static final String UPDATE_CHECK_PRERELEASE_OPTION = "Automatically check for GAPID pre-release updates";
```
"Include preview releases"

hevrard replied:
done

pmuetschard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -137,6 +141,9 @@ protected void configureDialog(FileDialog dialog) {
           createCheckbox(this, Messages.UPDATE_CHECK_OPTION, true),
           withSpans(new GridData(SWT.LEFT, SWT.TOP, false, false), 2, 1));
+      allowUpdateCheckPrereleases = withLayoutData(
+          createCheckbox(this, Messages.UPDATE_CHECK_PRERELEASE_OPTION, false),
```
Indent this checkbox, so it's clear that it's a dependent option.

hevrard replied:
done

pmuetschard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -108,6 +111,7 @@ private void update() {
     private final Button allowAnalytics;
     private final Button allowCrashReports;
     private final Button allowUpdateChecks;
+    private final Button allowUpdateCheckPrereleases;
```
maybe find a shorter name for this?

hevrard replied:
also used "includePrereleases"

pmuetschard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -156,6 +163,14 @@ protected void configureDialog(FileDialog dialog) {
       adbPath.addBoxListener(SWT.Modify, adbListener);
       adbListener.handleEvent(null);
+
+      SelectionListener allowUpdateChecksListener = new SelectionAdapter() {
```
```
allowUpdateChecks.addListener(SWT.Selection,
    e -> allowUpdateCheckPrereleases.setEnabled(allowUpdateChecks.getSelection()));
```

Also consider clearing the checkbox if the auto-update checkbox is cleared.

hevrard replied:
done

hevrard commented:
Thanks @pmuetschard for the feedback, I addressed the comments.

Now I would like GAPIC to call `watchForUpdates()` when the update settings are modified, or maybe more easily whenever the settings dialog is closed. Yet, watchForUpdate needs a GAPIS Client, etc, which is available in the MainWindow, but not in Settings.

What would be the best way to trigger an update check if people change the Update settings?

hevrard commented on core/app/run.go:
```diff
@@ -97,6 +97,11 @@ func (v VersionSpec) GreaterThan(o VersionSpec) bool {
 	}
 }
 
+// Equal returns true if v is equal to o, ignoring the Build field
```
done

hevrard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -34,6 +34,9 @@
 import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.events.SelectionListener;
```
removed

hevrard commented on gapic/src/main/com/google/gapid/views/SettingsDialog.java:
```diff
@@ -156,6 +163,14 @@ protected void configureDialog(FileDialog dialog) {
       adbListener.handleEvent(null);
+
+      allowUpdateChecks.addListener(SWT.Selection,
+          e -> {
```
fixed

hevrard commented on gapis/server/server.go:
```diff
@@ -129,13 +129,29 @@ func (s *server) CheckForUpdates(ctx context.Context, includePrereleases bool) (
+
+		// dev-releases are previews of the next release, so e.g. 1.2.3 is more recent than 1.2.3-dev-456
+		if version.GreaterThan(mostRecentVersion) ||
+			(version.Equal(mostRecentVersion) && (devVersion == -1 || devVersion > mostRecentDevVersion)) {
```
Yes indeed, thanks! Fixed.

hevrard commented:
Fix #3065
